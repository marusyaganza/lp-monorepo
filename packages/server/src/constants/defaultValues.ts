import { Language, Role, SortWordsBy, Tense } from '../generated/graphql';

export const WORDS_PER_PAGE = 5;
export const DEFAULT_LANGUAGE = Language.English;
export const DEFAULT_WORDS_SORT_BY = SortWordsBy.UpdatedAt;
export const DEFAULT_ROLE = Role.Member;
export const DEFAULT_TENSE = Tense.Pind;
export const DEFAULT_TOKEN_TTL = 7;
export const DEFAULT_MONGO_CONNECTION = 'mongodb://localhost:27017';
export const DEFAULT_OPTIONS__SAMPLE_SIZE = 50;
