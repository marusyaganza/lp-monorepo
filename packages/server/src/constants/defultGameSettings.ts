import { Game, Language, SettingsByGame } from '../generated/graphql';

export const DEFAULT_GAMES_SETTINGS: SettingsByGame = {
  [Game.Audio]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Image]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Speaking]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.TypeWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectDef]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Conjugation]: {
    wordsPerGame: 2,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  },
  [Game.Gender]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  }
};
