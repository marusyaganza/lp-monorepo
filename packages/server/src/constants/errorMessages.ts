export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
