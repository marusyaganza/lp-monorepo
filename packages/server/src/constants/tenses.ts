import { Tense } from '../generated/graphql';

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto',
  futr: 'Futuro indicativo',
  cond: 'Condicional',
  psub: 'Presente subjentivo'
  // gppt: 'Gerundio'
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
