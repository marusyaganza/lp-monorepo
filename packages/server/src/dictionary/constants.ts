import { Tense } from 'generated/graphql';

export type TagType = {
  tag: string;
  replacement: string;
};

export type ComplexTagType = {
  opening: string;
  closing: string;
};

export const TAGS: TagType[] = [
  { tag: '{b}', replacement: '<b>' },
  { tag: '{/b}', replacement: '</b>' },
  { tag: '{ldquo}', replacement: '\u201C' },
  { tag: '{rdquo}', replacement: '\u201D' },
  { tag: '{it}', replacement: '<i>' },
  { tag: '{/it}', replacement: '</i>' },
  { tag: '{wi}', replacement: '<i>' },
  { tag: '{/wi}', replacement: '</i>' },
  { tag: '{gloss}=', replacement: '<i>' },
  { tag: '{/gloss}', replacement: '</i>' },
  { tag: '{sc}', replacement: '<i>' },
  { tag: '{/sc}', replacement: '</i>' },
  { tag: '{parahw}', replacement: '</i>' },
  { tag: '{/parahw}', replacement: '</i>' },
  { tag: '{phrase}', replacement: '</i>' },
  { tag: '{/phrase}', replacement: '</i>' },
  { tag: '{qword}', replacement: '</i>' },
  { tag: '{/qword}', replacement: '</i>' },
  { tag: '{itsc}', replacement: '<i>' },
  { tag: '{/itsc}', replacement: '</i>' },
  { tag: '{inf}', replacement: '' },
  { tag: '{/inf}', replacement: '' },
  { tag: '{sup}', replacement: '' },
  { tag: '{/sup}', replacement: '' },
  { tag: '{bc}', replacement: '' },
  { tag: '{dx}', replacement: '' },
  { tag: '{/dx}', replacement: '' },
  { tag: '{dx_def}', replacement: '' },
  { tag: '{/dx_def}', replacement: '' },
  { tag: '{dx_ety}', replacement: '' },
  { tag: '{/dx_ety}', replacement: '' },
  { tag: '{ma}', replacement: '' },
  { tag: '{/ma}', replacement: '' },
  { tag: '{rom}', replacement: '' },
  { tag: '{/rom}', replacement: '' }
];

export const COMPLEX_TAGS: ComplexTagType[] = [
  { opening: '{sx', closing: '}' },
  { opening: '{d_link', closing: '}' },
  { opening: '{a_link', closing: '}' },
  { opening: '{i_link', closing: '}' },
  { opening: '{et_link', closing: '}' },
  { opening: '{mat', closing: '}' },
  { opening: '{dxt', closing: '}' }
];

export const COMPLEX_TAGS_TO_REMOVE: ComplexTagType[] = [
  { opening: '{ds|', closing: '}' }
];

export const HW_TAGS: TagType[] = [
  { tag: '{bit}', replacement: '' },
  { tag: '{/bit}', replacement: '' }
];

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto'
  //   futr: 'Futuro indicativo',
  //   cond: 'Condicional',
  //   psub: 'Presente subjentivo',
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
