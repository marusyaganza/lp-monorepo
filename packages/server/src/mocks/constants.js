export const mockEnv = {
  SALT_NUMBER: 3,
  JWT_SECTET: 'JWT_secret'
};
