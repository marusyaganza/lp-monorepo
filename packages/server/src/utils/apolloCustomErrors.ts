import { GraphQLError } from 'graphql';
import { ApolloServerErrorCode } from '@apollo/server/errors';
import { ERROR_MESSAGES } from '../constants/errorMessages';

export enum ERROR_CODES {
  Unauthorized = 'UNAUTHORIZED',
  Forbidden = 'FORBISSEN'
}

export class AuthenticationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHENTICATED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Unauthorized } });
  }
}

export class AuthorizationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHORIZED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Forbidden } });
  }
}

export class UserInputError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: { code: ApolloServerErrorCode.BAD_USER_INPUT }
    });
  }
}

export class OperationResolutionError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: {
        code: ApolloServerErrorCode.OPERATION_RESOLUTION_FAILURE
      }
    });
  }
}
