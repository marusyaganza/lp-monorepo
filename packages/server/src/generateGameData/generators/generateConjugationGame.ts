import {
  Game,
  GameData,
  GameDataInput,
  GameQuestion,
  Word
} from '../../generated/graphql';
import { getGameTask } from '../helpers';
import { OperationResolutionError } from '../../utils/apolloCustomErrors';
import { DEFAULT_TENSE } from '../../constants/defaultValues';
import { ERROR_MESSAGES } from '../../constants/errorMessages';

export async function generateConjugationGame(
  words: Word[],
  parameters: GameDataInput
): Promise<GameData> {
  const gameType = Game.Conjugation;
  const tense = parameters?.tense || DEFAULT_TENSE;

  const questions: GameQuestion[] = words
    .map(word => {
      const { imgUrl, shortDef, audioUrl, name, id, conjugation } = word;
      const verbForms = conjugation?.find(conj => conj?.cjid === tense)?.cjfs;

      if (!verbForms) {
        throw new OperationResolutionError(
          ERROR_MESSAGES.GAME_GENERATION_FAILED
        );
      }

      return {
        question: [name],
        answer: verbForms.join(', '),
        wordId: id,
        additionalInfo: {
          imgUrl,
          audioUrl,
          shortDef: `<b>${name} means</b> ${shortDef[0]}`
        }
      };
    })
    .filter(Boolean);

  const task = getGameTask(gameType, tense);

  return { questions, task, type: gameType, tense };
}
