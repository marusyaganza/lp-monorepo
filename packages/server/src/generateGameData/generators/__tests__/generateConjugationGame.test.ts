import {
  wordsWithConjugation,
  testData
} from '../../../tests/mocks/wordsForGames';
import { generateConjugationGame } from '../generateConjugationGame';
import { Game, Language, Tense } from '../../../generated/graphql';
import { TENSES } from '../../../constants/tenses';
import { OperationResolutionError } from '../../../utils/apolloCustomErrors';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';
import { DEFAULT_GAMES_SETTINGS } from '../../../constants/defaultGameSettings';

const gameType = Game.Conjugation;
const gameConfig = DEFAULT_GAMES_SETTINGS[gameType];

const language = Language.Spanish;
const tenses = Object.values(Tense);

describe('generateConjugationGame', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  tenses.forEach(tense => {
    it(`should return the correct data if there are enough words with ${tense} tense`, async () => {
      const gameData = await generateConjugationGame(wordsWithConjugation, {
        language,
        gameType,
        tense
      });
      expect(
        gameData.questions?.map(q => ({ ...q, wordId: expect.any(String) }))
      ).toMatchSnapshot();
      expect(gameData?.task).toEqual(`Conjugate the verb in ${TENSES[tense]}`);
      expect(gameData.type).toEqual(gameType);
      expect(gameData?.questions?.length).toBe(gameConfig?.wordsPerGame);
      expect(gameData.type).toBe(gameType);
    });

    it(`generate data with minimal ${language} words with ${tense} tense`, async () => {
      const words = wordsWithConjugation.slice(0, 1);
      const gameData = await generateConjugationGame(words, {
        language: language,
        gameType,
        tense
      });

      expect(gameData?.questions?.length).toBe(gameConfig?.minWords);
      expect(gameData.type).toBe(gameType);
      expect(gameData).toMatchSnapshot();
    });
    it('should throw an error with incorrect data', async () => {
      let error;
      try {
        // @ts-expect-error: using incorrect arguments here
        await generateConjugationGame(testData[language]);
      } catch (err) {
        error = err;
      }
      expect(error).toEqual(
        new OperationResolutionError(ERROR_MESSAGES.GAME_GENERATION_FAILED)
      );
    });
  });
});
