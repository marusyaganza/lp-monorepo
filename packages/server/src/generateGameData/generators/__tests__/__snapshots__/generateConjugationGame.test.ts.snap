// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateConjugationGame generate data with minimal SPANISH words with impf tense 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "audioUrl": "https://example.com/audio/pensar.mp3",
        "imgUrl": "https://example.com/images/pensar.jpg",
        "shortDef": "<b>pensar means</b> To think.",
      },
      "answer": "piensa, piense, pensemos, pensad, piensen",
      "question": [
        "pensar",
      ],
      "wordId": "001",
    },
  ],
  "task": "Conjugate the verb in Imperativo",
  "tense": "impf",
  "type": "CONJUGATION",
}
`;

exports[`generateConjugationGame generate data with minimal SPANISH words with pind tense 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "audioUrl": "https://example.com/audio/pensar.mp3",
        "imgUrl": "https://example.com/images/pensar.jpg",
        "shortDef": "<b>pensar means</b> To think.",
      },
      "answer": "pienso, piensas, piensa, pensamos, pensáis, piensan",
      "question": [
        "pensar",
      ],
      "wordId": "001",
    },
  ],
  "task": "Conjugate the verb in Presente indicativo",
  "tense": "pind",
  "type": "CONJUGATION",
}
`;

exports[`generateConjugationGame generate data with minimal SPANISH words with pprf tense 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "audioUrl": "https://example.com/audio/pensar.mp3",
        "imgUrl": "https://example.com/images/pensar.jpg",
        "shortDef": "<b>pensar means</b> To think.",
      },
      "answer": "he pensado, has pensado, ha pensado, hemos pensado, habéis pensado, han pensado",
      "question": [
        "pensar",
      ],
      "wordId": "001",
    },
  ],
  "task": "Conjugate the verb in Pretérito perfecto simple",
  "tense": "pprf",
  "type": "CONJUGATION",
}
`;

exports[`generateConjugationGame generate data with minimal SPANISH words with pret tense 1`] = `
{
  "questions": [
    {
      "additionalInfo": {
        "audioUrl": "https://example.com/audio/pensar.mp3",
        "imgUrl": "https://example.com/images/pensar.jpg",
        "shortDef": "<b>pensar means</b> To think.",
      },
      "answer": "pensé, pensaste, pensó, pensamos, pensasteis, pensaron",
      "question": [
        "pensar",
      ],
      "wordId": "001",
    },
  ],
  "task": "Conjugate the verb in Pretérito imperfecto",
  "tense": "pret",
  "type": "CONJUGATION",
}
`;

exports[`generateConjugationGame should return the correct data if there are enough words with impf tense 1`] = `
[
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/pensar.mp3",
      "imgUrl": "https://example.com/images/pensar.jpg",
      "shortDef": "<b>pensar means</b> To think.",
    },
    "answer": "piensa, piense, pensemos, pensad, piensen",
    "question": [
      "pensar",
    ],
    "wordId": Any<String>,
  },
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/correr.mp3",
      "imgUrl": "https://example.com/images/correr.jpg",
      "shortDef": "<b>correr means</b> To run.",
    },
    "answer": "corre, corra, corramos, corred, corran",
    "question": [
      "correr",
    ],
    "wordId": Any<String>,
  },
]
`;

exports[`generateConjugationGame should return the correct data if there are enough words with pind tense 1`] = `
[
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/pensar.mp3",
      "imgUrl": "https://example.com/images/pensar.jpg",
      "shortDef": "<b>pensar means</b> To think.",
    },
    "answer": "pienso, piensas, piensa, pensamos, pensáis, piensan",
    "question": [
      "pensar",
    ],
    "wordId": Any<String>,
  },
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/correr.mp3",
      "imgUrl": "https://example.com/images/correr.jpg",
      "shortDef": "<b>correr means</b> To run.",
    },
    "answer": "corro, corres, corre, corremos, corréis, corren",
    "question": [
      "correr",
    ],
    "wordId": Any<String>,
  },
]
`;

exports[`generateConjugationGame should return the correct data if there are enough words with pprf tense 1`] = `
[
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/pensar.mp3",
      "imgUrl": "https://example.com/images/pensar.jpg",
      "shortDef": "<b>pensar means</b> To think.",
    },
    "answer": "he pensado, has pensado, ha pensado, hemos pensado, habéis pensado, han pensado",
    "question": [
      "pensar",
    ],
    "wordId": Any<String>,
  },
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/correr.mp3",
      "imgUrl": "https://example.com/images/correr.jpg",
      "shortDef": "<b>correr means</b> To run.",
    },
    "answer": "he corrido, has corrido, ha corrido, hemos corrido, habéis corrido, han corrido",
    "question": [
      "correr",
    ],
    "wordId": Any<String>,
  },
]
`;

exports[`generateConjugationGame should return the correct data if there are enough words with pret tense 1`] = `
[
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/pensar.mp3",
      "imgUrl": "https://example.com/images/pensar.jpg",
      "shortDef": "<b>pensar means</b> To think.",
    },
    "answer": "pensé, pensaste, pensó, pensamos, pensasteis, pensaron",
    "question": [
      "pensar",
    ],
    "wordId": Any<String>,
  },
  {
    "additionalInfo": {
      "audioUrl": "https://example.com/audio/correr.mp3",
      "imgUrl": "https://example.com/images/correr.jpg",
      "shortDef": "<b>correr means</b> To run.",
    },
    "answer": "corrí, corriste, corrió, corrimos, corristeis, corrieron",
    "question": [
      "correr",
    ],
    "wordId": Any<String>,
  },
]
`;
